File: 2025-03-10/Diameter_Of_Binary_Tree.java
import java.util.ArrayList;

public class Diameter_Of_Binary_Tree {
    private static int ans;
    public static int solve(TreeNode root) {
        if(root == null) return 0;

        int l = solve(root.left);
        int r = solve(root.right);

        ans = Math.max(ans, l + r);

        return 1 + Math.max(l, r);
    }
    public static int diameterOfBinaryTree(TreeNode root) {
        ans = 0;
        solve(root);
        return ans;
    }

    public static void main(String[] args) {
        Integer[] arr = new Integer[]{3,1,null,null,2};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        int ans = diameterOfBinaryTree(root);
        System.out.println(ans);
    }
}

File: 2026-02-13/Diameter_Of_Binary_Tree.py
from typing import Optional, List
from Tree import TreeNode, Tree


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxLen = [0]
        def getLen(node):
            if not node:
                return 0

            left = getLen(node.left)
            right = getLen(node.right)
            maxLen[0] = max(maxLen[0], 1 + left + right)

            return 1 + max(left, right)

        getLen(root)
        return maxLen[0]-1


sol = Solution()
tree = Tree()
node = tree.array_to_tree([1,2,3,4,5])
print(sol.diameterOfBinaryTree(node))
